### ISSUE TITLE
To Serialize or not to Serialize

### ISSUE THREAD

**Issue by lwallent** [Romakita/ts-express-decorators#137]
Hi,

I wish to have my Swagger UI generated by TsED and store my data using Mongoose.

> In the future I would love to generate my mongoose schema from my model as well...

Currently I have a serialisation issue that I would like to discuss here? And perhaps get some idea on how to proceed?

I will create a model and a controller for storing the following Example:

```typescript
export interface IExample {
    simpleProperty: string;
    complexProperty: {
        someProperty:number;
    }
}
``` 

In order to make this model show in Swagger I create the following two classes:

```typescript
export class Complex {
    @JsonProperty()
    someProperty: number;
}

export class Example implements IExample {  
    @JsonProperty()
    simpleProperty:string;

    @JsonProperty()
    complexProperty: Complex;
}
```

To store this we create a simple mongoose schema (should later be generated from POJOs)

```typescript
const schemaDefinition = {
    simpleProperty: {type: String},

    complexProperty: {      
        someProperty: {type: Number}
    }
}

const exampleSchema: Schema = new Schema(schemaDefinition);

export type ExampleDocument = IExample & Document;
export const ExampleStore = mongoose.model<ExampleDocument>('Example', exampleSchema);
```
We add a simple controller to allow saving of the examples:

```typescript
@Controller("/examples")
export class ExampleCtrl {

    @Post("/")
    @Returns(200, {use: Example})
    async create(@Required() @Name("example")  @BodyParams() example: Example): Promise<Example> {
        const storable = new ExampleStore(example);
        await storable.validate();
        return  await storable.save();
    }
}
```
**NOTE: The above will unfortunately not save the complex property**

The problem is the additional class needed to decorate the nested property in order to get the wanted Swagger UI! The serialisation done by TsED is correct but not really wanted!? To save using mongoose we need the lean representation.

To make it save correctly one could do something crazy like the following:

```typescript
    async create(@Required() @Name("example") @Description("The example to create") @BodyParams() example: Example): Promise<Example> {
        const storable = new ExampleStore(JSON.parse(JSON.stringify(example)));
        await storable.validate();
        return  await storable.save();
    }
```

The data would also have the correct form if we were to skip the converters all together adding the following in BodyParams:

```typescript
 useConverter: false
```

but then the types would not be correct which would be undesirable.

What do you guys do in this case? Any ideas? Should we perhaps find a way to separate serialisation from swagger specification. Or, am I completely missing something here?

One could stick with an interface for the complexProperty and then make a hint for the swagger implementation - sort of:

```typescript
export class Example implements IExample {  
    @JsonProperty()
    simpleProperty:string;

    @JsonProperty()
    @Swagger({use: Complex})
    complexProperty: IComplex;
}
```

Help and suggestions are more than welcome. I would love to implement changes on both sides of the fence I just need to know where we should go on this

  Kind Regards 

Lasse

**Comment by Romakita** [comment]
Hi @lwallent ,

I'm not sure if I correctly understand your problem. I need an example ;)

What is the object structure sent to the method ?
```json
{
  ... replace the structure
}
```
What is the object structure when converter deserialize the object ?
```json
{
...replace the structure
}
```
See you ;)
What do you want as a result? 
```
{
...replace the structure
}
```

For me BodyParams() and Swagger must use the same Model, because it's the contract of service that you expose.

**Comment by lwallent** [comment]
Hi @Romakita,

Well I guess my problem can be reduced to the presence of classes as parameters in the endpoints - normally I would just have interfaces, but generating the Swagger requires the classes.

The problem is more related to Mongoose. I should probably investigate how Mongoose reads its data from objects. The issue can be reduced to the following **not** working with Mongoose;

```typescript
  let ex = new Example();
  ex.simpleProperty = "test";
  let cplx = new Complex();
  cplx.someProperty = 42;
  ex.complexProperty = cplx;

  const storable = new ExampleStore(ex);
  await storable.validate();
  return  await storable.save();
```
In this case only the following is saved (complex is missing):

````
{   "__v": 0,   "simpleProperty": "test",   "_id": "59d9573d229391a03c0b9369" }
````

Whereas the following does work:

```typescript
 let ex = {
   simpleProperty: "test",
   complexProperty: {
      someProperty: 42
   }
 }

 const storable = new ExampleStore(ex);
 await storable.validate();
 return  await storable.save();
```

Problem is that TsED deserilisation produces the first (which is type wise correct).

This might be a mongoose thing I was just wondering if you had experienced this?

Thanks for taking your time :-)

**Comment by Romakita** [comment]
Hi @lwallent ,

I've found a solution :)

In your MongooseService you can write this:
```typescript
import * as Mongoose from "mongoose";
const Document: Type<any> = (Mongoose as any).Document;

@Service()
export class MongooseService {
    static resource: Mongoose.Connection;

    /**
     *
     * @param {ConverterService} converterService
     */
    constructor(protected converterService: ConverterService) {

        MongooseService.override("set", (obj, ...args: any[]) => {
            return [
                converterService.serialize(obj),
                ...args
            ];
        });

        // optional.
        Document.prototype.toClass = function (target) {
            return converterService.deserialize(this.toObject(), target);
        };
    }
}
```

Example:
```typescript
async save(@BodyParams() event: CalendarEvent)  {
        // CalendarEventsService.checkPrecondition(event);
        $log.debug({message: "Validate event", event});
        const eventModel = new CalendarEventModel(event);

        await eventModel.validate();
        $log.debug({message: "Save event", eventModel});
        await eventModel.update(event, {upsert: true});

        event = eventModel.toClass(CalendarEvent);

        $log.debug({message: "Event saved", event, isEventClass: event instanceof CalendarEvent});
        return event;
}
```

I let you try this experimental solution ;). If it work fine for you, maybe we can write a PR in the documentation ?

See you,
Romain

**Comment by lwallent** [comment]
Hi @Romakita,

Thanks for the suggested solution! That is kind of a crazy approach :-)

I will go forward with it, however, I do not think personally that this should be added "as is" to the documentation. Rather I think the code it self should be added. I have given the whole mongoose thing some thought lately and I think we should rather make a separate folder (like Swagger) that will hold code for making _TsED_ play nice with Mongoose . This should include code for generating the Mongoose Schema and could also hold special code like the above, so that the complexity is hidden from the day-to-day use :-)

I will have a look at it when some time comes my way! Please let me know if you are already working on some Mongoose stuff, then I will not interfere :-)

Kind Regards and thanks for taking your time 

Lasse

**Comment by Romakita** [comment]
Hi @lwallent,

Yes I've this feature in mind (socket.io too), but nothing produced code. You have typegoose as example, work fine, but it's little tricky. (also node decorators).

It's a great idea if you work on ;). I can create the story in github and write some example :).
See you,
Romain

**Comment by lwallent** [comment]
Hi @Romakita,

I worked with typegoose for some time (their code as well) but in the end I backed away as I disliked the way my model classes had to extend the Typegoose class. I would like to have completely independent model classes and then just use the decorated classes to generate the mongooseSchema (complete separation).

You are welcome to add a separate issue/story where the details could be discussed. I can describe my take on it as well :-)

Kind Ragards

Lasse
